### ISSUE TITLE
Add a ThrowsAsync method to return a failed task

### ISSUE THREAD

**Issue by thomaslevesque** [FakeItEasy/FakeItEasy#932]
While thinking about how [automatic cancellation](https://github.com/FakeItEasy/FakeItEasy/issues/904) should work for async methods, I realized that the current behavior for `Throws` isn't ideal for these methods. The problem is that there are two ways for a `Task`-returning method to "throw":

1. it can throw an exception directly, if its implementation is not actually asynchronous but just has a `Task` return type. A typical case is a method that just returns the result of another async method, e.g. in order to validate the arguments synchronously:

        public Task<Foo> LoadFooAsync(string path)
        {
            // the exception will be thrown immediately when the method is called
            if (path == null) throw new ArgumentNullException(nameof(path));
            return LoadFooAsyncImpl(path);
        }

        private async Task<Foo> LoadFooAsyncImpl(string path)
        {
            using (var stream = File.OpenRead(path))
            using (var reader = new StreamReader(stream))
            {
                string text = reader.ReadToEndAsync();
                return Foo.Parse(text);
            }
        }

2. or it can return a failed task, if its implementation really is asynchronous:

        public async Task<Foo> LoadFooAsync(string path)
        {
            // the exception will NOT be thrown immediately when the method is called,
            // but only when the task is awaited
            if (path == null) throw new ArgumentNullException(nameof(path));

            using (var stream = File.OpenRead(path))
            using (var reader = new StreamReader(stream))
            {
                string text = reader.ReadToEndAsync();
                return Foo.Parse(text);
            }
        }

The difference doesn't really matter in most cases, because if you directly `await` the result of the method, the behavior will be the same. But if instead you want to do something with the resulting task, it makes a big difference, because in the first case the method will throw directly and you won't get a task at all. For instance, if you do something like this:
```
LoadFooAsync(path).ContinueWith(...);
```
if `path` is null, `LoadFooAsync` will throw and `ContinueWith` won't be executed. Of course, when you have async/await, using `ContinueWith` becomes less common, but there are other scenarios where you would want to manipulate the tasks directly, e.g. parallel processing:
```
var fooTask = GetFooAsync();
var barTask = GetBarAsync();

await Task.WhenAll(fooTask, barTask);
```
In cases like that, the difference between throwing directly and returning a failed task becomes important.

OK, that was a long introduction, but I'm getting to the point :wink:

Currently, FakeItEasy only offers option 1 : configure a method to throw an exception directly. This is fine in most cases, but won't work for scenarios when the task needs to be manipulated, rather than just awaited. Of course, it's possible to explicitly configure the method to return a failed task, but it's not very convenient:
```
A.CallTo(() => fake.GetFooAsync()).Returns(Task.FromException<Foo>(new Exception("oops")));
```
(also, `FromException` is only available since .NET 4.6)

FIE already provides a shortcut to return a value directly from an async method, rather than explicitly configure it to return a completed task. We could have a similar shortcut for returning a failed task:
```
A.CallTo(() => fake.GetFooAsync()).ThrowsAsync(new Exception("oops"));
```
The name clearly indicates that it "throws asynchronously", rather than directly throw an exception.

The only drawback I can see is that it would be slightly misleading, since `ThrowsAsync` would actually configure a return value.

**Comment by adamralph** [comment]
> The only drawback I can see is that it would be slightly misleading, since `ThrowsAsync` would actually configure a return value.

I don't think this is a problem. In the world of async, `Task` is the new `void`, and `Task<int>` is the new `int`.

**Comment by blairconrad** [comment]
Darn. I'd meant to comment on this a while ago. Sorry, @thomaslevesque.

For me, `ThrowsAsync` is figureoutable, but not obvious. I worry that it could be misinterpreted as a kind of "*Lazily"
Also, one drawback (I think) is that it will be available on every method with a return value, not just the ones that return tasks, no?
So we'll have to catch misuses and throw exceptions at runtime?

**Comment by thomaslevesque** [comment]
> Also, one drawback (I think) is that it will be available on every method with a return value, not just the ones that return tasks, no?

Unless it's implemented as an extension method. But I'm afraid we're going to hit the same issue as in #804, where some extension methods suddenly had an extra generic type parameter, making them hard to use because both generic type parameters had to be specified (which is why we eventually put them in the interface)

**Comment by thomaslevesque** [comment]
> But I'm afraid we're going to hit the same issue as in #804

Actually, it works fine in the basic cases:

    public static void ThrowsAsync(this IReturnValueConfiguration<Task> configuration, Exception ex)
    {
        configuration.ReturnsLazily(call => Task.FromException(ex));
    }

    public static void ThrowsAsync<T>(this IReturnValueConfiguration<Task<T>> configuration, Exception ex)
    {
        configuration.ReturnsLazily(call => Task.FromException<T>(ex));
    }

    public static void ThrowsAsync(this IReturnValueConfiguration<Task> configuration, Func<IFakeObjectCall, Exception> exceptionFactory)
    {
        configuration.ReturnsLazily(call => Task.FromException(exceptionFactory(call)));
    }

    public static void ThrowsAsync<T>(this IReturnValueConfiguration<Task<T>> configuration, Func<IFakeObjectCall, Exception> exceptionFactory)
    {
        configuration.ReturnsLazily(call => Task.FromException<T>(exceptionFactory(call)));
    }

And for overloads like these:

    public static void ThrowsAsync<T, T1>(this IReturnValueConfiguration<Task<T>> configuration, Func<T1, Exception> exceptionFactory)
    {
        configuration.ReturnsLazily((T1 arg1) => Task.FromException<T>(exceptionFactory(arg1)));
    }

    public static void ThrowsAsync<T, T1, T2>(this IReturnValueConfiguration<Task<T>> configuration, Func<T1, T2, Exception> exceptionFactory)
    {
        configuration.ReturnsLazily((T1 arg1, T2 arg2) => Task.FromException<T>(exceptionFactory(arg1, arg2)));
    }

But not for overloads that take the exception type as a generic param:

    public static void ThrowsAsync<T, TException>(this IReturnValueConfiguration<Task<T>> configuration)
        where TException : Exception, new()
    {
        configuration.ReturnsLazily(call => Task.FromException<T>(new TException()));
    }

because in this case we need to specify both the exception type and the return type, as in `.ThrowsAsync<int, InvalidOperationException>()`. If C# allowed extension methods in generic classes, it would solve the issue, but unfortunately it's not the case...

I think not including the problematic overloads is an acceptable limitation.

**Comment by blairconrad** [comment]
> I think not including the problematic overloads is an acceptable limitation.

So, to recap, you can implement an acceptable subset of the desired methods that only apply to faked methods that return a task, and your only roadblock is my complaint about the name?
If that's the case, complaint withdrawn!

**Comment by thomaslevesque** [comment]
> So, to recap, you can implement an acceptable subset of the desired methods that only apply to faked methods that return a task, and your only roadblock is my complaint about the name?

Basically, yes 😉 

I don't mind picking another name, but I can't think of anything more adequate. `ReturnsFailedTask` would be accurate, but doesn't seem very natural to me.

**Comment by blairconrad** [comment]
I probably would've come around to `ThrowsAsync` even it was available when configuring any method.
Since it's limited to Task-returning methods, my concerns have essentially been nullified.
:shipit:

**Comment by thomaslevesque** [comment]
> And for overloads like these:
> ```
> public static void ThrowsAsync<T, T1>(this IReturnValueConfiguration<Task<T>> configuration, Func<T1, Exception> exceptionFactory)
> {
>     configuration.ReturnsLazily((T1 arg1) => Task.FromException<T>(exceptionFactory(arg1)));
> }
> 
> public static void ThrowsAsync<T, T1, T2>(this IReturnValueConfiguration<Task<T>> configuration, Func<T1, T2, Exception> exceptionFactory)
> {
>     configuration.ReturnsLazily((T1 arg1, T2 arg2) => Task.FromException<T>(exceptionFactory(arg1, arg2)));
> }
> ```

Actually, those extension methods have the same problem we had discussed in https://github.com/FakeItEasy/FakeItEasy/pull/804#discussion-diff-71817665

They work if we do this:

    A.CallTo(() => foo.BarAsync(A<int>._, A<string>._)).ThrowsAsync((int i, string s) => ...);

But if the generic arguments are specified explicitly, the return type has to be specified as well:

    A.CallTo(() => foo.BarAsync(A<int>._, A<string>._)).ThrowsAsync<double, int, string>((i, s) => ...);

(assuming `BarAsync` returns a `Task<double>`

It's not as bad as it was for `Throws` and a `AssignsOutAndRefParameter`, since it's just the return type that needs to be specified rather than the configuration interface, but it's still not a very nice experience... To solve this in #804 I had to move some of the extension methods to the interface. I could do the same here, although it's not ideal.

**Comment by thomaslevesque** [comment]
> I could do the same here, although it's not ideal.

Scratch that... these methods are specific to task-returning methods, so I can't put them in the general interface.

**Comment by thomaslevesque** [comment]
In fact, we already have the same problem with `ReturnsLazily` for async methods, and I had never noticed it before, so maybe it's not such a big deal...

This works:

    A.CallTo(() => foo.BarAsync(A<int>._, A<string>._)).ReturnsLazily((int i, string s) => ...);

But here we need to specify the return type:

    A.CallTo(() => foo.BarAsync(A<int>._, A<string>._)).ReturnsLazily<double, int, string>((i, s) => ...);

Anyway, I'll submit a PR soon, then we can discuss the pros and cons.

**Comment by blairconrad** [comment]
This issue has been released in FakeItEasy 3.0.0:
https://github.com/FakeItEasy/FakeItEasy/releases/3.0.0
